src/dev/lupio/lupio_blk.cc

```diff
@@ -0,0 +1,241 @@
+#define SECTOR_SIZE   (1ULL << SECTOR_BITS)
+
+namespace gem5
+{
+LupioBLK::LupioBLK(const Params &params) :
+    DmaDevice(params),
+    platform(*params.platform),
+    dmaEvent([this]{ dmaEventDone(); }, name()),
+    pioAddr(params.pio_addr),
+    pioSize(params.pio_size),
+    image(*params.image)
+{
+    nbBlocks = image.size();
+    DPRINTF(LupioBLK, "LupioBLK initalized\n");
+}
+
+static inline int clz64(uint64_t val)
+{
+    return val ? __builtin_clzll(val) : 64;
+}
+
+void
+LupioBLK::dmaEventDone()
+{
+    int64_t _offset = lba;
+
+    if (writeOp) {
+        for (int i = 0; i < reqLen; i += SECTOR_SIZE){
+            image.write(&reqData[i], _offset);
+            _offset++;
+        }
+    }
+    delete[] reqData;
+    busy = false;
+    DPRINTF(LupioBLK, "Done with DMA event\n");
+    platform.postPciInt(LUPIO_BLK_INT_ID);
+}
+
+uint64_t
+LupioBLK::lupioBLKRead(const uint8_t addr)
+{
+    uint64_t r = 0;
+
+    switch (addr >> 2) {
+        case LUPIO_BLK_CONF:
+            r = SECTOR_BITS << 16               /* log2(512B / block) */
+                | (63 - clz64(nbBlocks));      /* log2(Nb of blocks) */
+
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_CONF: %d\n", r);
+            break;
+        case LUPIO_BLK_NBLK:
+            r = nblk;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_NBLK: %d\n", r);
+            break;
+        case LUPIO_BLK_BLKA:
+            r = lba;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_BLKA: %d\n", r);
+            break;
+        case LUPIO_BLK_MEMA:
+            r = mem;
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_MEMA: %d\n", r);
+            break;
+
+        case LUPIO_BLK_STAT:
+            r = busy;                      /* Always idle */
+            if (writeOp) {
+                r |= LUPIO_BLK_TYPE;    /* Write command */
+            }
+            if (err) {
+                r |= LUPIO_BLK_ERRR;    /* Error */
+            }
+            DPRINTF(LupioBLK, "Read LUPIO_BLK_STAT: %d\n", r);
+
+            /* Acknowledge IRQ */
+            platform.clearPciInt(LUPIO_BLK_INT_ID);
+            break;
+
+        default:
+            panic("Unexpected read to the LupioRTC device at address %#llx!",
+                    addr);
+            break;
+    }
+    return r;
+}
+
+void
+LupioBLK::lupioBLKWrite(const uint8_t addr, uint64_t val64)
+{
+    uint32_t val = val64;
+    switch (addr >> 2) {
+        case LUPIO_BLK_NBLK:
+            nblk = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_NBLK: %d\n", nblk);
+            break;
+        case LUPIO_BLK_BLKA:
+            lba = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_BLKA: %d\n", lba);
+            break;
+        case LUPIO_BLK_MEMA:
+            mem = val;
+            DPRINTF(LupioBLK, "Write LUPIO_BLK_MEMA: %d\n", mem);
+            break;
+
+        case LUPIO_BLK_CTRL:
+            err = false;
+            writeOp = val & LUPIO_BLK_TYPE;
+
+            /* Perform command */
+            if (!busy) {
```

Comment:
This `!busy` needs to encompass the whole case statement (or you could even panic if the device is busy since technically no one should do that, and if they do then it's probably a software bug that needs fixing).